java/core/src/java/org/apache/orc/OrcFile.java

```diff
@@ -642,4 +671,187 @@ public static Writer createWriter(Path path,
+   * The user metadata are merged and files that disagree on the value
+   * associated with a key will be rejected.
+   *
+   * @param outputPath the output file
+   * @param options the options for writing with although the options related
+   *                to the input files' encodings are overridden
+   * @param inputFiles the list of files to merge
+   * @return the list of files that were successfully merged
+   * @throws IOException
+   */
+  public static List<Path> mergeFiles(Path outputPath,
+                                      WriterOptions options,
+                                      Path... inputFiles) throws IOException {
+    Writer output = null;
+    byte[] buffer = new byte[0];
+    Configuration conf = options.getConfiguration();
+    TypeDescription schema = null;
+    CompressionKind compression = null;
+    int bufferSize = 0;
+    Version fileVersion = null;
+    WriterVersion writerVersion = null;
+    int rowIndexStride = 0;
+    List<Path> result = new ArrayList<>(inputFiles.length);
+    Map<String, ByteBuffer> userMetadata = new HashMap<>();
+
+    for (Path input: inputFiles) {
+      FileSystem fs = input.getFileSystem(conf);
+      Reader reader = createReader(input,
+          readerOptions(options.getConfiguration()).filesystem(fs));
+
+      if (!understandFormat(input, reader)) {
+        continue;
+      } else if (schema == null) {
+        // if this is the first file that we are including, grab the values
+        schema = reader.getSchema();
+        compression = reader.getCompressionKind();
+        bufferSize = reader.getCompressionSize();
+        rowIndexStride = reader.getRowIndexStride();
+        fileVersion = reader.getFileVersion();
+        writerVersion = reader.getWriterVersion();
+        options.blockSize(bufferSize)
+            .version(fileVersion)
+            .writerVersion(writerVersion)
+            .compress(compression)
+            .rowIndexStride(rowIndexStride)
+            .setSchema(schema);
+        if (compression != CompressionKind.NONE) {
+          options.enforceBufferSize().bufferSize(bufferSize);
+        }
+        mergeMetadata(userMetadata, reader);
+        output = createWriter(outputPath, options);
+      } else if (!readerIsCompatible(schema, fileVersion, writerVersion,
+          bufferSize, rowIndexStride, compression, userMetadata, input,
+          reader)) {
+        continue;
+      } else {
+        mergeMetadata(userMetadata, reader);
+      }
+      List<OrcProto.StripeStatistics> statList =
+          reader.getOrcProtoStripeStatistics();
+      FSDataInputStream inputStream = fs.open(input);
+      int stripeNum = 0;
+      result.add(input);
+
+      for(StripeInformation stripe: reader.getStripes()) {
+        int length = (int) stripe.getLength();
+        if (buffer.length < length) {
+          buffer = new byte[length];
+        }
+        long offset = stripe.getOffset();
+        inputStream.readFully(offset, buffer, 0, length);
```

Comment:
The current WriterImpl.appendStripe() will fail if it is called on files that don't have stripe statistic, so that feels like we can handle it in a follow up bug.

If the merge can't complete, throwing an exception seems like the best course. I guess we can delete the partially written output file if that feels cleaner to you.